src/WidgetBase.ts

```diff
@@ -337,70 +270,20 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> extends E
 		}
 	}
 
-	public __render__(): VirtualDomNode | VirtualDomNode[] {
+	public __render__(): DNode | DNode[] {
 		this._renderState = WidgetRenderState.RENDER;
-		if (this._dirty === true || this._cachedVNode === undefined) {
+		if (this._dirty || this._cachedDNode === undefined) {
 			this._dirty = false;
 			const render = this._runBeforeRenders();
 			let dNode = render();
-			dNode = this.runAfterRenders(dNode);
-			this._decorateNodes(dNode);
-			const widget = this._dNodeToVNode(dNode);
-			this._manageDetachedChildren();
+			this._cachedDNode = this.runAfterRenders(dNode);
 			this._nodeHandler.clear();
-			this._cachedVNode = widget;
-			this._renderState = WidgetRenderState.IDLE;
-			return widget;
 		}
 		this._renderState = WidgetRenderState.IDLE;
-		return this._cachedVNode;
-	}
-
-	private _decorateNodes(node: DNode | DNode[]) {
-		let nodes = Array.isArray(node) ? [ ...node ] : [ node ];
-
-		this._numRootNodes = nodes.length;
-		this._currentRootNode =  0;
-		const rootNodes: DNode[] = [];
-		this._rootNodeKeys = [];
-
-		nodes.forEach(node => {
-			if (isHNode(node)) {
-				rootNodes.push(node);
-				node.properties = node.properties || {};
-				if (node.properties.key) {
-					this._rootNodeKeys.push(node.properties.key);
-				}
-			}
-		});
-
-		while (nodes.length) {
-			const node = nodes.pop();
-			if (isHNode(node) || isWNode(node)) {
-				node.properties = node.properties || {};
-				if (isHNode(node)) {
-					if (rootNodes.indexOf(node) !== -1 || node.properties.key) {
-						node.properties.afterCreate = this._afterCreateCallback;
-						node.properties.afterUpdate = this._afterUpdateCallback;
-					}
-					if (node.properties.bind === undefined) {
-						(<any> node.properties).bind = this;
-					}
-				}
-				else {
-					if (!node.coreProperties) {
-						node.coreProperties = {
-							bind: this,
-							baseRegistry: this._coreProperties.baseRegistry
-						};
-					}
-				}
-				nodes = [ ...nodes, ...node.children ];
-			}
-		}
+		return this._cachedDNode;
 	}
 
-	protected invalidate(): void {
+	public invalidate(): void {
```

Comment:
why the change is visibility?

Reply:
It now get's called from outside of the widget instance (in vdom), previously it was called from within the instance in `dNodeToVNode`

Reply:
👍